### ISSUE TITLE
'Block all HTTP requests' should attempt HTTPS

### ISSUE THREAD

**Issue by Manouchehri** [EFForg/https-everywhere#2115]
Lately I've been using the 'Block all HTTP requests' feature in Chrome, and for quite a few smaller sites (e.g., personal blogs) I find myself manually changing the URL to `https://`. It works for a surprising amount of sites.

Wouldn't it make sense to attempt to switch a HTTP URL to HTTPS, even if there's no defined rule already? Since the site is already blocked at this point, I can't see any downside aside from a few extra milliseconds by trying HTTPS as a last-ditch effort.

Another benefit is it's possible that some SysAdmins will see a bunch of failed connections over 443/TCP in their firewall logs and hopefully get the message that their users want HTTPS.

My apologizes if this has already been brought up, I did glance at the issue tracker and didn't see it mentioned.

**Comment by jsha** [comment]
This is a decent idea. We would only want to do it for top-level site navigation, I think. It requires a little careful thought on how the UI should work. If HTTPS also fails, we would want the user to wind up still on the HTTP page, with a message saying it was blocked for being HTTP. We don't want this feature to simply wind up showing users cert errors more often.

Also, there are some risks. Some sites (notably Forbes) serve different content on HTTPS vs HTTP. Also, I think many sites have an HTTPS version, but it is broken due to mixed content blocking. If we send users to the broken HTTPS version automatically, it makes things look like HTTPS Everywhere is broken rather than the site itself.

**Comment by Manouchehri** [comment]
I definitely think that this should be implemented as a non-default option, because you're right, it will break several sites silently in the process. To help make it clear, maybe add an indicator to show that no proper rules were found to apply, and HTTPS is being blindly applied.

I'm not sure if I'm following the reason behind only doing it for top-level URLs. What difference would it make?

True, I noticed that Forbes is sending a broken certificate as well (ironically, when I went to an article about Akamai serving invalid certificates, [I got an invalid certificate from Akamai](https://i.imgur.com/rSJRNXg.png)...).

**Comment by chaccaronmacarondo** [comment]
Yeah firefox was having that problem almost every website with any sort of Multimedia modules/trackers. 

For a while, the then-current Firefox wouldn't implement the feature you discuss without the aid of HTTPS everywhere.... Two new bespoke alpha builds later.... it was fixed.. 

You could set the requirement for OCSP chains of trust, from their original issuing authority by default in Firefox... Now that was useful.... Thus, if the request for stapling the ocsp chain failed in Firefox, the HTTPS Everywhere is usually able to follow the chain-of-trust now. 

 So, at one point, this watchlist for SSL observatory was working perfectly with a similar setting element enabled in 'about:config';    
    I think it was something like extension.HTTPSEverywhere.orig.ocsp.GET (for the second layer     of protection).

It only evr has been giving me problems, when connecting to sites hosted by ,"akamaized-staging.net, ... , etc." CertificateAuthorites (CA's).

Lastly, when it was working, HTTPS everywhere would subvert the warning in the Firefox taskbar, and, effectively send all the media requests through the website's server, of course with the ability to disable the https tracking rules for various services/adverts. 

So, in cases where ad-trackers' rules were enabled  (at places like Youtube),  these sorta sessions/connections got opened without any sort of 'hiccups', obviously., and would've still display the content, through the ad-trackers' rulesets by default.... 

 But now, the problem is back; and, on sites with "insecure images/content" the blocking is broken again...  It's very disconcerting... All i wanted to do was have firefox do dns requests routed through ipv4, so I could leave the (NAT) Network Adresss Translation service alone; cuz my bro needs matchmaking/conferencing on his Xbox... I can never set the router wirelessly, either, (for security's sake).

The fact that, where 'Block all HTTP requests' was disabled, the browser didn't  have any problems, or warnings, (the "(!)" symbol in the address bar on Firefox) displaying that nice little lock icon in the address bar. Now.., it's broken again... VERY disconcerting.

it should still be parsing the insecure elements in the webpage... 

Again, that's the culprit. Very annoying bc setting ipv4 IP protocol settings under the hood, now causes that nice toggle for HTTPS Everywhere to "staple" the original oscp's, to vanish... 

P.S. [sorry, screenshot uploaded from linux is broken]

**Comment by Manouchehri** [comment]
@chaccaronmacarondo I'm not sure if I'm able to follow what you're saying. Those sound like application issues, not plugin related. I don't understand how IPv4 relates to any of this; are you trying to talk about a SOCKS proxy possibly?

The only HTTPS Everywhere related topic I can see is that you're saying that HTTP requests are still going through on mixed content pages when they should be blocked instead, correct?

**Comment by rugk** [comment]
@jsha 
You argue against your own FAQ? [Why use a whitelist of sites that support HTTPS? Why can't you try to use HTTPS for every last site, and only fall back to HTTP if it isn't available?] (https://www.eff.org/https-everywhere/faq#faq-Why-use-a-whitelist-of-sites-that-support-HTTPS?-Why-can%27t-you-try-to-use-HTTPS-for-every-last-site,-and-only-fall-back-to-HTTP-if-it-isn%27t-available?)

And because of the there mentioned reasons I also think this is not good to do this - at least not in normal 'mode'. But this issue is about the 'Block all HTTP requests' feature and there this can of course be attempted and in case a HTTPS request fails HTTPS Everywhere just blocks this request. So instead of a fallback to something unsecure (HTTP) this would be a try to load the content in a secure way and if it fails it will be blocked.

**Comment by Manouchehri** [comment]
None of us are proposing to this by default, only as a secondary option to 'Block all HTTP requests' mode. This additional makes no sense on it's own; it's logical when you're already blocking HTTP though, what do you have to lose?

@rugk I don't believe @jsha was suggesting that we actually connect to the HTTP page, only *forward* back to it so the user gets the proper `This webpage was blocked by an extension` message instead of `This webpage is not available`.

**Comment by rugk** [comment]
> @rugk I don't believe @jsha was suggesting that we actually connect to the HTTP page, only forward back to it so the user gets the proper This webpage was blocked by an extension message instead of This webpage is not available. It's more of a UX point than a technical one.

Okay, in this it's all right. :smiley:

**Comment by 1tischmann1** [comment]
@Manouchehri  (this is, 'CHACCARONMACARONDO' -- I entered the wrong email for the signup thingy ,~~_me to hoon paagal_~~)

Just as  jsha, mentioned I guess it was the fact that it wasn't blocking all HTTP requests,  so the issue about the ipv4 dns lookups (in conjunction with HTTPS everywhere addon),  is moot, according to him.

In conjunction with this, I'm now saying that it'd be __inefficent__ to have a lookup function/table to parse all other plain-HTTP requests on the page, as these can be complicated; because, A myriad of unimportant things are called through HTTP-proper, (such as thumbnail images, CSS style sheets foreign ad-tracking apps, blablabla, _inter alia_)-- obviously VERY complicated-- the usefulness of chains of trust and ocsp, is therefore quite apparent. That seems to be the promise of HTTPS Everywhere application. 

 Somehow the settings I used caused HTTPS Everywhere, in " HTTP/HTTPS rules allowed " mode, might have affected the tooltip about insecure content, in Firefox. Ironically, it changed to a warning; although, the website, YouTube, was still able to be browsed safely. 

If it is  the HTTP requests that are broken because of the way they're processsed in your app, the concept we mentioned, "somehow" may very well contribute to the cause of hapazard warnings, security issues, in Firefox, which, in the context of User Experience, is disconcerting.

There are more than enough default warnings in Firefox,  e.g. about phishing; eavesdropping, etc.  to scare inexperienced users, like me, unwary of the fact that the downloading of some forms of HTTP are benign,  who are trusting this Firefox Application for what it does, and the placebo effect of safeguards for web-browsing, as well as the SSL-observatory thing... 


Someone must fix that problem with authenticating/aquiring  domains with 'Akamized' CAs...
(though, I trust not a problem for you to worry about). They really don't seems safe, at all.... The sheer number of all their guises for DNS lookups, is really rediculous. Not even Chinese CAs would implement something just as stupid... jeez :-[

@jsha @rugk

**Comment by 1tischmann1** [comment]
@Manouchehri  @1tischmann1 ;"  *bump** " ,please 

__bump``__  ,plz

@jsha  @rugk

**Comment by Manouchehri** [comment]
@1tischmann1 I'm sorry, I honestly can't follow what you're saying. I don't think this has anything to do with the original topic though. There's no need to bump the thread, we all already got notifications about the original post (it's also the weekend in every timezone).

HTTPS Everywhere already is able to parse everything and the efficiency seems fine to all of us. I'm not sure how else you could be changing HTTP to HTTPS to begin with without checking every page. Please read through the [FAQ](https://www.eff.org/https-everywhere/faq) to understand what this project is.

We can't force Akamai to start serving proper certificates. I'm still not sure how any of this is supposed to tie into DNS.

**Comment by 1tischmann1** [comment]
I think you might do well to narrow the scope of your answer. Although, I get your point in the second paragraph of your answer.

Nevertheless, regarding the idea about changing the apps behaviour for HTTP, as one op mentioned... that was one thing that I thought would be sensible to mention. Good ideas come in strange places, so why not make further mention that the feature request from @Manouchehri 
(oh wait, that was you!) either, might be "useful", or, post that just as a reminder that something like that might be too tedious, or inefficient to be implemented properly. I agree about breaking sites-- http links abound.

My comment on the Akamai CAs was a bit of a jest. A bit of irony, too, since I've actually never come across mention of this complaint about Akamai anywhere _at all_ ,save for the complaints that I read on this page. I saw some irony in the fact that, browsers such as Firefox already have this up-to-date functionality, making use of some new elements in this standard (by this, I mean Digital Certification and ocsp stapling, etc.).

It interested me to note that they've got some issues  with compliance, there-- Compliance is a bit of a rough idea, in that regard, but the standards /should/ tend to change as new tech is being developed.   ~~I need not mention: how LONG these broken Cert IDs are gonna be around, considering how many people probably ignore some aspects of browsing security and add an exception, anyway, causing the host "akamai...blabla" to be exempt from blocking over secure connections.~~ A travesty, I'm sure.

 To say the least, Bugs/glitches can get very complicated. Sorry, again for my indignation and rudeness. More likely than not, things might be taken-down a bit easier if you're one of those  Russian Security-analysts (Kapersky)... So "rest easy, man", heh

As for the issue I would like to bring up, I've described (in layman's terms) more than once, with acute specificity, a little bit of this quirk, that I would like to talk about (no pun intended). 

 The app was useful, and reassuring, at first, because setting "block HTTP requests" would get rid of that pesky warning. But that setting isn't the point, as it isn't as pertinent for general cases of:

...particular search-result links on Google, (or even some unencrypted websites, for which accessing is 'disallowed', under that rule) as opposed to the use of some proper HTTPS redirects on a site like DuckDuckGo, to make a simple comparison. (although they sometimes fail, when linking to sites that don't do HTTPS.

 As pertains to one example of a "generality", which is very common:

E.g.~ Sites like Imdb.com still don't do some specific types of HTTPS encryption; although they do say they are going to newly implement HTTPS connections as a feature (or, maybe they already have).


In general, using HTTPS everywhere with standard settings, as well as with some rules disabled [again, not the important bit of the issue, for me,] is presenting me with a bit of an inconsistency in one such case. (_caveat_ ,(be aware): As it is well known, some websites have mixed, https and html elements.)


Again, I shall say that the issue is not superficial. At some point the warning went away after I changed some settings. I'm guessing, that MOST of the requests in the environment of the web page, such as it was (both https and http hyperlinks on youtube) should be being sent/processed through HTTPS Everywhere at some point, because this would involve the behaviour of calling/ignoring requests to load data from http sources, with HTTPS Everywhere still set to **default**.

Symptomatically, the conditions, under which the issue I'm experiencing have occured, still have to do with your addon processing requests for webpages. The culprit lay where the behaviour of the Addon is involved, at an interface where normal requests **actually** aren't being parsed by your App. versus the default (no loopholes) behaviour of browser security warnings that are present in Firefox; warnings which may present themselves, during any non-encrypted ["unsecure"] connections that might put the end-user at risk. By default, this warning was what Firefox automatically deigned to report to me, the User (even using https everywhere in the 'default' manner, on a safe site), by way of a "warning message" in the Address Bar. 

Obviously YouTube script on the Browser was sill requesting to lookup HTTP links,  _even_ with your app enabled--- Indeed, with standard settings, HTTPS everywhere should not (and, rather effectually, _doesn't_) block requests for http connections. Instead, it only applies the proper Rules; and/or SSL Observatory checks where connections through SSL are allowed. 

Somehow, the behaviour of the browser was such that, with regard to automatically using the features of HTTPS everywhere Addon, the alert never appeared again, for a time, after changing some settings in 'about:config'. That was comforting, lol.

Also, some settings have been newly implemented in Firefox, to the extent of including optional switches encompassing the (recent) accordance with new ''ocsp stapling requests''.

[[ I.e. ,the option to _always_ send 'Orig.ocsp.GET.require' requests in conjunction with the /other/ options for ocsp chain-stapling. ]]


Somehow, with **default** settings for your app, at one point the warning symbol had not appeared, for a time, and now, because I **set** conditions causing an inconsistency in 'about:config' I can't actually reproduce this behaviour.

Exactly what I was saying before; I liked it when I **didn't** need to use the HTTP nowhere mode to get rid of that superficial warning. I feel like this needs to be covered/evaluated. The only settings, and these alone, that were changed to see the difference in this neat behaviour, were:

-IPv6 settings;
-Behaviours for GET/Request.orig.ocsp [true/enabled]; 
-That one little ocsp.GET rule/entry that comes with HTTPS everywhere; 
And, the most recent screwball change had to do with DNS prefetching settings. Idk what could be the actual culprit.


Jeez, I wish I knew how to speak Wolof, it would make it sooo much easier to scold people who aren't being in the positive...

**Comment by rugk** [comment]
@1tischmann1
Your 'TL;DR' is longer than everything else so I doubt that many people will read this text.
Can't you just summarize this *a bit*? And put the things which does not have to do with the original issue in a new/other issues?

**Comment by 1tischmann1** [comment]
well, I only relate this issue here because I first noticed the thing Manouchehri mentioned, about the behaviour of block all http requests mode. The point is, Firefox, sometimes doesn't send requests for unsecure elements on a webpage. HTTPS everywhere adds a layer of protection by enforcing rule-sets (among other things). 

Specifically, now, which is what I’ve said in every single paragraph of that longer blurb, the behaviour of HTTPS Everywhere, that is of interest to me, is the fact that Firefox would automatically display the exclamation ("!") symbol about unsecure elements on the page.

Here's the issue: 


In effect, what Manoucheri has already mentioned here--- in order to properly connect to the websites that are broken in this manner, using 'block http requests' mode, the user sometimes must type in the url again.

Notwithstanding the way the browser normally behaves, even with HTTPS everywhere enabled, this behaviour occurred at a "safe" website (YouTube), regardless of whichever mode was selected in HTTPS Everywhere, Also, regardless of whichever rule-sets were enabled for various elements on the webpage.

Now, one peculiarity, is that the warning would normally periodically reappear while browsing youtube pages, unless I closed the session/connection to YouTube, by refreshing the page. After changing the default ocsp handling in Firefox about:config, somehow, the warning (which was usually present while browsing this website, even with the app) was no longer showing up. 

**DONE.** (I hope this is of interest, here vv, Read On!)

Just recently, I reset 'about:config to defaults, as well as deleting some of the current SSL tokens, Sig.'s, and some irrelevant Software Security Devices; this was in order to clear up some clutter. so I was able to re-authenticate my connections at certain websites. 

(This notion, hopefully, won't start you Open Source-er's hesds spinning round, though, because it IS rather tricky to describe, in its entirety.)

It seems like, somewhere, the way that HTTPS Everywhere handles the request/response pipeline for SSL connections, doesn't evaluate, in rather important situations, whether-or-not Firefox is **set** to use incompatible handling/behaviour.

It was interesting, because somehow after refreshing firefox, and cleaning out outdated tokens, for sites I needn't visit anymore, another strange thing occured... HTTPS Everywhere, for some odd reason, started showing in its menu, **duplicate** rulesets  (i.e. a ruleset, on one page, appearing to have loaded **twice**, from just one host).

^^That last behaviour is where the issue becomes a bit more than a simple feature-request, as requested by OP, but the peculiarity of this, sort  of 'smacked' in a very similar manner, when I read this discussion. @rugk  @jsha @Manouchehri

**Comment by rugk** [comment]
The thing you talk about in the first part is the [Mixed content blocking] (https://www.eff.org/https-everywhere/rulesets#mixed-content) problem of HTTPSE.
So I just summarized this in one sentence.

The second thing seems to be a different issue. AFAIK HTTPSE has no effect on the validation of TLS connection. Firefox or Chrome validate the connection like they would always do.
However the thing with duplicate rulesets displayed in the tray menu has nothing to do with this issue as far as I see it so if you want to report this, please open a new issue.

**Comment by Manouchehri** [comment]
Thanks!
